### PR TITLE
GitHub actions CI/CD releases

### LINKED ISSUE #6: Provide GitHub release binaries / artifacts via CI
I'm not sure the exact details but I know this can be done. This'll simplify a lot of things for those who don't have rust installed.

### PR DESCRIPTION
**Pull request by praveenperera**
Adds CI and CD using GitHub releases. Closes #6 

Will create a new release when a new `v*` tag is pushed. You can see it here: 

https://github.com/praveenperera/bore/tree/github-actions-ci-cd

### CODE REVIEW

**Comment by ekzhang** [comment]
Wow, thanks for the contribution! This is great. Let me take a look and get back to you on this. There's almost as many lines of code in CI as there are for the rest of the Rust codebase. :')

This is based on https://github.com/japaric/trust, right?

**Comment by praveenperera** [comment]
Yup! Its based on trust and https://github.com/XAMPPRocky/mean-bean-ci-template

I'm also using it in:

- https://github.com/cantino/mcfly
- https://github.com/avencera/rustywind

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -30,7 +30,7 @@ jobs:
           toolchain: stable
           components: rustfmt
 
-      - run: cargo fmt -- --check
+      - run: rustfmt --edition 2021 **/*.rs --check
```
Comment (ekzhang) [inline .github/workflows/ci.yml:33]: Why use `rustfmt` directly instead of `cargo fmt`?

File: .github/workflows/mean_bean_ci.yml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        platform: [linux-musl]
+
+  macos:
+    runs-on: macos-latest
+    strategy:
+      fail-fast: true
+      matrix:
+        channel: [stable, beta, nightly]
+        target:
+          - x86_64-apple-darwin
```
Comment (ekzhang) [inline .github/workflows/mean_bean_ci.yml:36]: That's a lot of CI runs! I'm glad GitHub Actions is free for public repositories. :')

**Review by ekzhang — APPROVED** [review]
Read through the scripts, looks good to me. I'm a little bit cautious about recommending that people pipe a bash script downloaded from the internet to `sudo sh`, but I can't think of any better solution than this.

I'll try running CI on your branch, update the README a bit (there's a new "installation" section to update now), and if all looks good will merge.

**Comment by ekzhang** [comment]
Hm, I think Windows is missing from the CI and deploy steps! Right now it's only building for Linux and macOS.

Also `bore` only needs to be compiled on `stable`, we don't support `beta` or `nightly`.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (praveenperera) [reply .github/workflows/ci.yml:33]: Hey I copied it in from a different repo but ya we can use either here.

**Comment by praveenperera** [comment]
@ekzhang Added windows and removed beta and nightly

**Comment by praveenperera** [comment]
@ekzhang looks like a test is failing on windows: https://github.com/praveenperera/bore/runs/5983694419?check_suite_focus=true

**Comment by ekzhang** [comment]
Ah, I see, guess the tests don't work on Windows. The problem is likely `TcpListener::bind("localhost:0")`. I'm surprised that this was never an issue for anyone though.

**Comment by ekzhang** [comment]
Going to release 0.3.0 for now since there's a few pending features, but I'll try to debug the Windows problem as soon as I get the chance.

**Comment by ekzhang** [comment]
This looks fantastic! The build looks like it works perfectly and the binaries appear to work as well. Do you think it's okay if we remove `install.sh` though? There's a couple concerns I have:

1. People on non-Unix environments will get errors if they try to move things to `/usr/local/bin`. They also might not necessarily have `bash` or `curl` installed.
2. Asking people to `curl ... | sudo sh -s` something from a GitHub URL is somewhat scary and can open up supply chain attack vectors.

I could add something like this in `README.md`:

---

```shell
# Installation (requires Rust, see alternatives below)
cargo install bore-cli

# On your local machine
bore local 8000 --to bore.pub
```

(... then, further down)

## Installation

The easiest way to install `bore` is from prebuilt binaries. These are available on the [releases page](https://github.com/ekzhang/bore/releases) for macOS, Windows, and Linux. Just unzip the appropriate file for your platform and move the `bore` executable into a folder on your PATH.

You also can build `bore` from source using [Cargo](https://doc.rust-lang.org/cargo/), the Rust package manager. This command installs the `bore` binary at a user-accessible path.

```shell
cargo install bore-cli
```

---

**Comment by praveenperera** [comment]
Hey @ekzhang, I personally don't agree with sudo install scripts are scary. But both sides have been discussed a lot in other forums. But basically comes down to people are comfortable with package managers, and the bash script it not any more safe or dangerous than a package manager. Also anyone that is afraid can look at the install script right in the repo before installing. It's not too big or complicated. 

That being said I understand the fear, and this is your repo, I have removed it.

**Pull request merged**